src/target/sam4l.c

```diff
@@ -0,0 +1,410 @@
+#define	FLASHCALW_FCR				(FLASHCALW_BASE + 0x00)
+#define	FLASHALCW_FCR_WS1OPT		(1 << 7)
+#define	FLASHALCW_FCR_FWS			(1 << 6)
+#define	FLASHALCW_FCR_ECCE			(1 << 4)
+#define	FLASHALCW_FCR_PROGE			(1 << 3)
+#define	FLASHALCW_FCR_LOCKE			(1 << 2)
+#define	FLASHALCW_FCR_FRDY			(1 << 0)
+
+/* Flash Command Register */
+#define	FLASHCALW_FCMD				(FLASHCALW_BASE + 0x04)
+#define FLASHCALW_FCMD_KEY_MASK		0xff
+#define FLASHCALW_FCMD_KEY_SHIFT	24
+#define FLASHCALW_FCMD_PAGEN_MASK	0xffff
+#define FLASHCALW_FCMD_PAGEN_SHIFT	8
+#define FLASHCALW_FCMD_CMD_MASK		0x3f
+#define FLASHCALW_FCMD_CMD_SHIFT	0
+
+#define FLASH_CMD_NOP		0
+#define FLASH_CMD_WP		1	/* Write Page */
+#define FLASH_CMD_EP		2	/* Erase Page */
+#define FLASH_CMD_CPB		3	/* Clear Page Buffer */
+#define FLASH_CMD_LP		4	/* Lock page region */
+#define FLASH_CMD_UP		5	/* Unlock page region */
+#define FLASH_CMD_EA		6	/* Erase All */
+#define FLASH_CMD_WGPB		7	/* Write General Purpose Fuse Bit */
+#define FLASH_CMD_EGPB		8	/* Erase General Purpose Fuse Bit */
+#define FLASH_CMD_SSB		9	/* Set Security Fuses */
+#define FLASH_CMD_PGPFB		10	/* Program General Purpose Fuse Byte */
+#define FLASH_CMD_EAGPF		11	/* Erase All GP Fuses */
+#define FLASH_CMD_QPR		12	/* Quick Page Read (erase check) */
+#define FLASH_CMD_WUP		13	/* Write User Page */
+#define FLASH_CMD_EUP		14	/* Erase User Page */
+#define FLASH_CMD_QPRUP		15	/* Quick Page Read User Page */
+#define FLASH_CMD_HSEN		16	/* High Speed Enable */
+#define FLASH_CMD_HSDIS		17	/* High Speed Disable */
+
+/* Flash Status Register */
+#define FLASHCALW_FSR				(FLASHCALW_BASE + 0x08)
+#define FLASHCALW_FSR_LOCK(x)		(1 << (16 + (x)))
+#define FLASHCALW_FSR_ECCERR		(1 << 9)
+#define FLASHCALW_FSR_ECCERR2		(1 << 8)
+#define FLASHCALW_FSR_HSMODE		(1 << 6)
+#define FLASHCALW_FSR_QPRR			(1 << 5)
+#define FLASHCALW_FSR_SECURITY		(1 << 4)
+#define FLASHCALW_FSR_PROGE			(1 << 3)
+#define FLASHCALW_FSR_LOCKE			(1 << 2)
+#define FLASHCALW_FSR_FRDY			(1 << 0)
+
+/* Flash Parameter Register */
+#define FLASHCALW_FPR				(FLASHCALW_BASE + 0x0a)
+#define FLASHCALW_FPR_PSZ_MASK		0x7	/* page size */
+#define FLASHCALW_FPR_PSZ_SHIFT		8
+#define FLASHCALW_FPR_FSZ_MASK		0xf	/* flash size */
+#define FLASHCALW_FPR_FSZ_SHIFT		0
+
+/* Flash Version Register */
+#define FLASHCALW_FVR				(FLASHCALW_BASE + 0x10)
+#define FLASHCALW_FVR_VARIANT_MASK	0xf
+#define FLASHCALW_FVR_VARIANT_SHIFT	16
+#define FLASHCALW_FVR_VERSION_MASK	0xfff
+#define FLASHCALW_FVR_VERSION_SHIFT	0
+
+/* Flash General Purpose Registers (high) */
+#define FLASHCALW_FGPFRHI			(FLASHCALW_BASE + 0x14)
+/* Flash General Purpose Registers (low) */
+#define FLASHCALW_FGPFRLO			(FLASHCALW_BASE + 0x18)
+
+static void sam4l_extended_reset(target *t);
+static int sam4l_flash_erase(struct target_flash *f, target_addr addr, size_t len);
+static int sam4l_flash_write_buf(struct target_flash *f, target_addr dest,
+                             const void *src, size_t len);
+
+const struct command_s sam4l_cmds[] = {
```

Comment:
Do that in a separate PR when this one is merged if you want to.  It's optional.